test/support/spec_helpers/tmp.rb

```diff
@@ -0,0 +1,56 @@
+# Creates a temporary directory in the current working directory
+# for temporary files created while running the specs. All specs
+# should clean up any temporary files created so that the temp
+# directory is empty when the process exits.
+
+SPEC_TEMP_DIR_PID = Process.pid
+
+if spec_temp_dir = ENV["SPEC_TEMP_DIR"]
+  spec_temp_dir = File.realdirpath(spec_temp_dir)
+else
+  # TODO: revert when File.realpath is implemented
+  #spec_temp_dir = "#{File.realpath(Dir.pwd)}/rubyspec_temp/#{SPEC_TEMP_DIR_PID}"
+  spec_temp_dir = "./tmp.#{SPEC_TEMP_DIR_PID}"
+end
+SPEC_TEMP_DIR = spec_temp_dir
+
+SPEC_TEMP_UNIQUIFIER = "0"
+
+at_exit do
+  begin
+    if SPEC_TEMP_DIR_PID == Process.pid
+      # TODO: revert when Dir.delete is implemented
+      #Dir.delete SPEC_TEMP_DIR if File.directory? SPEC_TEMP_DIR
+      `rm -r #{SPEC_TEMP_DIR}` if File.directory? SPEC_TEMP_DIR
+    end
+  rescue SystemCallError
+    STDERR.puts <<-EOM
+
+-----------------------------------------------------
+The rubyspec temp directory is not empty. Ensure that
+all specs are cleaning up temporary files:
+  #{SPEC_TEMP_DIR}
+-----------------------------------------------------
+
+    EOM
+  rescue Object => e
+    STDERR.puts "failed to remove spec temp directory #{SPEC_TEMP_DIR}"
+    STDERR.puts e.message
+  end
+end
+
+def tmp(name, uniquify = true)
+  mkdir_p SPEC_TEMP_DIR unless Dir.exist? SPEC_TEMP_DIR
+
+  if uniquify and !name.empty?
+    # TODO: revert when String#rindex and String#insert are implemented
+    #slash = name.rindex "/"
+    #index = slash ? slash + 1 : 0
+    #name.insert index, "#{SPEC_TEMP_UNIQUIFIER.succ!}-"
+    slash = name.reverse.index("/")
+    index = slash ? name.size - slash - 1 : 0
+    name = name[0...index] + "#{SPEC_TEMP_UNIQUIFIER.succ!}-" + name[index..-1]
+  end
+
+  File.join SPEC_TEMP_DIR, name
+end
```

Comment:
D'oh! I didn't realize you were copying code from [mspec](https://github.com/ruby/mspec).

At the very least, we have to include a copyright header and a note in the README about such files copied.

Ideally, we would implement this stuff our selves without looking at mspec. I realize that's going to be very hard for you now that you've seen this file, so I'd accept option 1 above. Let's add a copyright notice to this file and the readme.

If you have copied any other files, we'll have to do the same.

I should have mentioned that we only copy files from `ruby/spec` and anything else beyond that, we need to discuss before we copy it. Even though they are the same MIT license, I want to be very careful not to claim ownership of something we didn't create.

Reply:
This is a rather misfortunate misunderstanding on my part.  I guess I had not mentally made the distinction between `ruby/spec` and `ruby/mspec`. I realize they are separate repos, but had considered that it was all part and parcel of the same thing.  Regarding the licensing, and the files themselves, for every file from ruby/spec copied in, no additional license header goes into the file included with Natalie, and I had assumed it was covered in the LICENSE.md at the project level and likewise any other "spec" stuff (fixtures, helpers, etc).

To further elaborate on the above, I was under the impression that the self-hosted "mini-mspec" was solely a matter of convenience to bootstrap the Natalie spec environment enough using the features that already existed in Natalie at the time, and once enough of Natalie was complete it would switch over to reference mspec directly.  I didn't get that a project goal was to build and maintain an independent testing environment for the long haul.  Perhaps that part also gets implemented in C++, dunno - I don't have a vision of the project goals.  I'm here because I want to learn more about Ruby internals and brush up on my poor C++ skills and hopefully help out a project at the same time.

In any case, I am interested in either: 

+ Rescinding my PR in its entirety if there is concern over this project being tainted.  
+ Reverting back to a commit prior to inclusion of mspec pieces if that is appropriate.
    + If you'd like any part of the contribution, i am happy to start over and submit piece by piece.
+ Keeping the code, and including the copyright notice in the appropriate place - _It was never my intention to not apply proper attribution_.  I am unclear if the best place would be in the `/test/support/spec_helpers` folder where the mspec files were brought in, in each file individually or at the project level.

Reply:
Let's put some copyright headers in the top of each file copied and call that done. I can do it if that would be helpful.

As for bootstrapping the spec suite until such time we can use mspec... That may be what we end up doing. We may indeed throw away our spec runner and use mspec some day. And in the light, I can certainly see how we would take shortcuts to get there!

But in any case, I think attribution is fine for this, maybe more fine than I implied in the earlier comment.

I do super appreciate all the work you've put into this PR and cannot wait to merge it!